### ISSUE TITLE
_app.getInitialProps and getStaticProps are being fired for all statically served files while deployed in production on Vercel 

### ISSUE THREAD

**Issue by jmikrut** [vercel/next.js#26085]
### What version of Next.js are you using?

10.2.0

### What version of Node.js are you using?

14.17.0

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

Vercel

### Describe the Bug

Hi all,

We have an app that uses the following NextJS features:

1. Custom `_app` with `getInitialProps` that fetches some global data from a headless CMS
2. A page template like `[...slug].tsx`
3. The page template above uses `getStaticProps` with `revalidate: XX` and `fallback: true`, and there are about 30 paths output by `getStaticPaths`. Each fetches its own data from the headless CMS

If we pop a console log in the `_app.getInitialProps` or in the `[...slug]`'s `getStaticProps`, we see that upon requesting the `[...slug].tsx` page, both the `_app.getInitialProps` AND `getStaticProps` run for _**all**_ static files that are served by NextJS - including things like `/favicon.ico` and files within the `/_next` directory.

This means that for every request to a page, when it revalidates, NextJS is making about 20+ fetches to the CMS and it's causing performance issues already. 

**This is only happening while being served in production on Vercel. It is not happening if serving locally.**

![image](https://user-images.githubusercontent.com/6651896/121920247-dc54af00-cd05-11eb-8cb0-4ad5fe3fde00.png)

### Expected Behavior

`getStaticProps` and `_app.getInitialProps` should not run for statically served files while being deployed on Vercel.

### To Reproduce

1. Create a new, blank NextJS project
2. Create a custom `_app` with `getInitialProps` and pop a console log in that function
3. Create a `[...slug].tsx` page template with `getStaticProps` using `revalidate: 10` and `fallback: true` - and then use `getStaticPaths`
4. Pop a console log in the `getStaticProps` function as well
5. Build
6. Serve in production
7. Watch Vercel's `View Function Logs` and see the console logs come in for static files (even `favicon.ico`) when you request a single page 

Sample repo:

https://github.com/trouble/vercel-bug

**Comment by sgtpsibin** [comment]
+1 same issue with getServerSideProps. 
_app.getInitialProps catch the server side request of getServerSideProps, cause _app.getInitialProps run twice.

**Comment by jmikrut** [comment]
I opened up a support ticket with Vercel on this one, for everyone's info. No response yet.

**Comment by mcsdevv** [comment]
Hi @jmikrut, I can see that you have been working with a member of the team to make progress on this since June 14th - are you referring to another case where you haven't received a response?

**Comment by jmikrut** [comment]
Hey @mcsdevv thank you for your response! I have been working with someone, yes, but last communication was on June 17 via email:

> Hi James

As you already know, NextJS is a complex framework. Your explanation is now clear to me and I can have reproduced the behavior in your repository. However, whether all static assets are loaded on revalidate or not and if yes, whether that can be controlled/configured is something that is beyond my ability to answer.

I will submit this as a question to the NextJS team and see what they have to say.

@mcsdevv feel free to follow up with me. I can provide the name of my contact and talk in further depth whenever!

**Comment by ijjk** [comment]
Hi, this behavior sounds expected as `getInitialProps` for `_app` is called during revalidation and a catch-all page is being used which can catch un-matched URLs like `favicon.ico`. If you don't want `favicon.ico` to match the catch-all page you can add a `favicon.ico` file to your `public` folder or add a rewrite to rewrite `/favicon.ico` -> `/404`. 

Related issue: https://github.com/vercel/next.js/issues/9819

I'm going to close this since this doesn't sound like a bug, if there is something else that doesn't seem correct that wasn't described above please reply with additional info and we can investigate further!

**Comment by jmikrut** [comment]
Hey @ijjk your suggested solution is not possible. 

If you look at the screenshot I posted of Vercel logs above, `getInitialProps` is running even for static JSON files in the `_next` folder—which I do not know the paths of to write redirects or rewrites to `404` manually. 

**`getInitialProps` should definitely not be running for those files.** I personally also think it should not be running for `favicon.ico` either, because that is literally never desirable. In general, it would be greatly preferred to simply not run `getInitialProps` ever for static files.

Does this clarify the issue?

**Comment by ijjk** [comment]
@jmikrut if revalidate is being used those `_next/data` requests are expected to be calling `_app`'s `getInitialProps` because it's also calling `getStaticProps` to return the data on a client-transition. Whenever a page with `getStaticProps` is revalidated the data route is also revalidated to keep them in sync.  

If you only want specific routes to be handled and no additional ones you can use `fallback: false` and add limit which paths you want to be handled in `getStaticPaths`.

**Comment by jmikrut** [comment]
@ijjk I think we might be talking about different things.

When I make a GET request, in Chrome, just loading a static file in the `_next` folder like this:

`/_next/data/i--C9ILuuPSFTVCZR7MxV/us/resources.json`

`getInitialProps` is fired.

My revalidation setting is 60 seconds, and even if I request this file more than once within that revalidation window, I can see in my server logs that `getInitialProps` is firing _each time_ I load that one file directly.

Why would `getInitialProps` be fired to retrieve a static file, within the revalidation window? It does not (at least should not) have anything to do with returning new data.

Just for even more context: I have 2 fetches to my CMS in `getInitialProps`. This effectively means that even though my site is statically rendered, every time someone requests a page (doesn't matter if within or outside the revalidation window), those two requests are fired _each_ like 6-10 times, all within a second of time. This cripples my CMS server. Even a $20 DigitalOcean droplet gets swamped with traffic.

I would be much better off server-rendering the entire page in this case if this is not a bug, as I would have ~1/10th of the requests to my CMS.

**Comment by jmikrut** [comment]
Also, one other note that I would like to reiterate:

The problems that I am reporting do not happen locally in production, nor do they happen when we hav a custom Express server. It only happens when deploying on Vercel.

**Comment by ijjk** [comment]
@jmikrut I think I see what's causing confusion. When `revalidate` is used for a page and the items are prerendered the initial request for the page will cause a revalidation on Vercel but sub-sequent visits will not until the revalidation period has elapsed. The users will always hit the cached version of the page if the path is prerendered via `getStaticPaths`.  For one revalidation on Vercel it is currently expected to see multiple log items in the function logs as we revalidate the HTML and the page data at the same time separately so that they can be kept in sync and cached separately for visiting the page path and the `_next/data` path. Do note the revalidate period is specific to each path so one path's revalidation period could have elapsed while another's has not. 

If you visit `/static/favicon.ico` as used in the reproduction it will not trigger the fallback or revalidation to occur as Next.js on Vercel does match static files before dynamic routes like the `[...slug].js` catch-all the same way Next.js does locally. If you were to not return `revalidate` for any paths in `getStaticProps` these would also not trigger `getInitialProps` or `getStaticProps` to be called after deploying as they would then be considered completely static. 

Using the above provided repo I could not reproduce `getStaticProps` or `getInitialProps` being called for each visit within the revalidate period besides the above noted initial requests done to cache the data and HTML separate for a path. 



https://user-images.githubusercontent.com/22380829/124057790-c3c6e300-d9ed-11eb-86c9-e0ce3a92e54c.mp4

**Comment by jmikrut** [comment]
Totally understand the intent behind how revalidation should work. We're on the same page there.

Thanks for pulling down my test repo to evaluate this. Try and load one of those static JSON files like I referenced above, though, instead of the page itself.

I just upped the revalidation window to 60 seconds in the demo repo, and I've redeployed and taken a screen recording myself. You can see that when loading a file from the `/_next/data` folder, even faster within the 60 second window, no matter how many times I load it, `getInitialProps` and `getStaticProps` are firing.

https://user-images.githubusercontent.com/6651896/124122500-28149180-da44-11eb-92ad-a9d5beab6652.mp4

Can you try loading a file in the `/_next` directory with your demo? The issue seems sproadic and strange, but the problem remains:

**If I want to run a statically generated site with ~60 second revalidation, currently Vercel swamps my CMS with way too many fetches to get data, and it would be more efficient for me to just run getServerSideProps**. 

So how do I fix this?

**Comment by ijjk** [comment]
Thanks for gathering that information and screen recordings, I've shared this internally with our edge team and we're taking a closer look at this and our team will update here when we have more information!

**Comment by jmikrut** [comment]
No problem. My team and I literally love Next and I will do everything I can to help here!

**Comment by jmikrut** [comment]
Hey @ijjk — any update here? I still see this happening on `12.0.9`.

**Comment by ijjk** [comment]
Reading back over the threads relating to this we haven't been able to reproduce static assets that exist causing revalidations. If a `_next` path doesn't exist and is requested e.g. `_next/non-existent` or `favicon.ico` and no `public/favicon.ico` file exists a catch-all route can match these unless `fallback: false` is used or is rewritten before matching the catch-all dynamic route. 

I also re-visited the initial shared reproduction https://vercel-bug-vert.vercel.app/another and checking the logs I'm not seeing non-stop revalidations within the revalidate period for a specific path as described. 

There have been some improvements to ISR on Vercel so it would be good to confirm this is still an issue with the latest version of Next.js v12.2.2.

**Comment by balazsorban44** [comment]
This issue has been automatically closed because it wasn't verified against next@canary. If you think it was closed by accident, please leave a comment. If you are running into a similar issue, please open a new issue with a reproduction. Thank you.
